lib/rmt.toit

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2022 Toitware ApS. All rights reserved.
+// Use of this source code is governed by an MIT-style license that can be
+// found in the lib/LICENSE file.
+
+import gpio
+
+/**
+Support for the ESP32 Remote Control (RMT).
+
+The $Channel represents a channel in the controller.
+
+An $Item represents an item from the controller.
+*/
+
+/**
+An Item to be transferred or received with RMT.
+
+An RMT item consists of a level (low or high) and a period (the amount of
+  ticks the level is sustained).
+
+# Advanced
+The period is specified in number of ticks, so the actual time the level is
+  sustained is determined by the RMT controller configuration.
+
+At the lower level, an item consits of 16 bits: 15 bits for the period and 1
+  bit for the level.
+*/
+class Item:
+  level/int
+  period/int
+
+  /** Constructs an item from the given $period and $level. */
+  constructor period level:
+    this.period = period & 0x7FFF
+    this.level = level & 0b1
+
+  /** Deserialized an item from the given $bytes at the given $index. */
+  constructor.from_bytes index/int bytes/ByteArray:
+    period = bytes[index] | ((bytes[index + 1] & 0x7F) << 8)
+    level = bytes[index + 1] >> 7
+
+  first_byte_ -> int:
+    return period & 0xFF
+
+  second_byte_ -> int:
+    return (period >> 8 ) | (level << 7)
+
+  /** See $super. */
+  operator == other/any:
+    if other is not Item: return false
+
+    return level == other.level and period == other.period
+
+  /** See $super. */
+  stringify -> string:
+    return "($period, $level)"
+
+/**
+An RMT channel.
+
+The channel must be configured after construction.
+
+  The channel can be configured for either RX or TX.
+*/
+class Channel:
+  num/int
+  pin/gpio.Pin
+
+  res_/ByteArray? := null
+
+  /** Constructs a channel using the given $num using the given $pin. */
+  constructor .pin .num:
+    res_ = rmt_use_ resource_group_ num
+
+  /**
+  Configure the channel for RX.
+
+  - $mem_block_num is the number of memory blocks (512 bytes) used by this channel.
+  - $clk_div is the source clock divider. Must be in the range [0,255].
+  - $flags is the configuration flags. See the ESP-IDF documentation for available flags.
+  - $idle_threshold is the amount of clock cycles the receiver will run without seeing an edge.
+  - $filter_en whether the filter is enabled.
+  - $filter_ticks_thresh pulses shorter than this value is filtered away.
+    Only works with $filter_en. The value must be in the range [0,255].
+  */
+  config_rx
```

Comment:
It might make sense to keep the configuration (and not apply it immediately), so that a channel can be used for read and writing.
The configuration could then be applied every time the direction switches.